drivers: dma: stm32: add null callback checking

ttwards opened the pull request:
drivers: dma: stm32: check for null dma_callback before enabling irqs

The STM32 DMA driver did not check if the optional dma_callback was set in dma_stm32_configure(). This could lead to a hard fault when an interrupt occurs and the callback is NULL, even though the API allows the callback to be omitted.

This patch modifies dma_stm32_configure() to enable transfer and half-transfer interrupts only if a valid dma_callback is supplied in config. This prevents NULL-function pointer dereference if callbacks are not used.

Fixes #97454

etienne-lms reviewed (commented):
Nitpicking on coding guidelines, otherwise imeplementation LGTM.

Regarding the commit message, I think you should rephrase the last sentence that does not really describe the change IMHO.
My suggestion:
```patch
 drivers: dma: stm32: add null callback checking
 
-The STM32 DMA driver did not check if the optional dma_callback was set
+Fix the STM32 DMA driver did not check if the optional dma_callback was set
 in dma_stm32_configure(). This could lead to a hard fault when an
 interrupt occurs and the callback is NULL, even though the API allows
 the callback to be omitted.

-This patch modifies dma_stm32_configure() to enable transfer and
-half-transfer interrupts only if a valid dma_callback is supplied in
-config. This prevents NULL-function pointer dereference if callbacks
-are not used.
-
 Fixes: ...
 Signed-off-by: ...

AUtech-dev reviewed (commented):
Thanks for the PR!
I can confirm that the current state fixes the hard fault.

However, I am a little bit concerned about the performance impact of the interrupts if the callback is not used:
The half transfer isr only clears the isr flag with `dma_stm32_clear_ht()` and does nothing more. Not requesting the interrupt at all (by not enabling it in `dma_stm32_configure`) would not change the system behavior but saves some cpu cycles - from my point of view.

In case of a cyclic dma tranfser (`stream->cyclic == true`) the transfer complete isr has the same behavior.

From my point of view this diff would improve performance:

```diff --git a/drivers/dma/dma_stm32.c b/drivers/dma/dma_stm32.c
--- a/drivers/dma/dma_stm32.c
+++ b/drivers/dma/dma_stm32.c
@@ -491,10 +501,13 @@
 	DMA_InitStruct.PeriphRequest = config->dma_slot;
 #endif
 	LL_DMA_Init(dma, dma_stm32_id_to_stream(id), &DMA_InitStruct);
 
-	LL_DMA_EnableIT_TC(dma, dma_stm32_id_to_stream(id));
+	/* Enable transfer complete isr if in non-cyclic mode or a callback is requested*/
+	if (!stream->cyclic || stream->dma_callback) {
+		LL_DMA_EnableIT_TC(dma, dma_stm32_id_to_stream(id));
+	}
 
-	/* Enable Half-Transfer irq if circular mode is enabled */
-	if (stream->cyclic) {
+	/* Enable Half-Transfer irq if circular mode is enabled and a callback is requested */
+	if (stream->cyclic && stream->dma_callback) {
 		LL_DMA_EnableIT_HT(dma, dma_stm32_id_to_stream(id));
 	}
```

Background: In my application I have a dma for RX and TX data with a quite fast timing and really need only one interrupt because both data paths are handled together.

ttwards committed:
drivers: dma: stm32: add null callback checking

Fix the STM32 DMA driver did not check if the optional dma_callback was
set in dma_stm32_configure(). This could lead to a hard fault when an
interrupt occurs and the callback is NULL, even though the API allows
the callback to be omitted.

Fixes #97454
Signed-off-by: Wenxi Xu <xuwenxi0517@gmail.com>

AUtech-dev reviewed (approved):
Tested this on hardware (stm32u073Xc). Works fine.

ttwards commented on drivers/dma/dma_stm32.c:
```diff
@@ -127,18 +128,25 @@ static void dma_stm32_irq_handler(const struct device *dev, uint32_t id)
-				     callback_arg, -EIO);
+		if (stream->dma_callback != NULL) {
+			stream->dma_callback(dev, stream->user_data,
+				    			 callback_arg, -EIO);
```
I understand, but two callings of `stream->dma_callback` reach 100~110 chars.
Shoud I use tab only?

AUtech-dev commented:
> > > From my point of view this diff would improve performance:
> >
> > @AUtech-dev While related, this is a different topic. Please raise a dedicated PR to discuss about it.
> 
> If your also convinced the proposed optimization is fully valid and valuable, it would deserve a dedicated commit. Unless what, as @erwango said, let's address this optimization in a later P-R.

I have opened https://github.com/zephyrproject-rtos/zephyr/pull/98057 for this optimization.

ttwards commented:
Updated. Sorry it's a bit busy for me nowadays. Thank you a lot!!!

etienne-lms commented:
@ttwards, you need to rebase tour patch.

mathieuchopstm commented:
rebase on behalf of @ttwards + address comments by @etienne-lms
